smart-contracts/contracts/CosmosBridge.sol

```diff
@@ -117,64 +129,73 @@ contract CosmosBridge is CosmosBridgeStorage {
+        if (usedNonce[_prophecyID]) {
+            claimComplete = Oracle(oracle).newOracleClaim(_prophecyID, msg.sender);
+        } else {
+            address tokenAddress;
+            string memory symbol;
+            if (_claimType == ClaimType.Burn) {
+                require(
+                    BridgeBank(bridgeBank).getLockedFunds(_symbol) >= _amount,
+                    "Not enough locked assets to complete the proposed prophecy"
+                );
+                symbol = _symbol;
+                tokenAddress = BridgeBank(bridgeBank).getLockedTokenAddress(_symbol);
+            } else if (_claimType == ClaimType.Lock) {
+                symbol = concat(COSMOS_NATIVE_ASSET_PREFIX, _symbol); // Add 'e' symbol prefix
+                address bridgeTokenAddress = BridgeBank(bridgeBank).getBridgeToken(symbol);
+                if (bridgeTokenAddress == address(0)) {
+                    // First lock of this asset, deploy new contract and get new symbol/token address
+                    tokenAddress = BridgeBank(bridgeBank).createNewBridgeToken(symbol);
+                } else {
+                    // Not the first lock of this asset, get existing symbol/token address
+                    tokenAddress = bridgeTokenAddress;
+                }
             } else {
-                // Not the first lock of this asset, get existing symbol/token address
-                tokenAddress = bridgeTokenAddress;
+                revert("Invalid claim type, only burn and lock are supported.");
             }
-        } else {
-            revert("Invalid claim type, only burn and lock are supported.");
-        }
 
-        // Create the new ProphecyClaim
-        ProphecyClaim memory prophecyClaim = ProphecyClaim(
-            _claimType,
-            _cosmosSender,
-            _ethereumReceiver,
-            msg.sender,
-            tokenAddress,
-            symbol,
-            _amount,
-            Status.Pending
-        );
+            // Create the new ProphecyClaim
+            ProphecyClaim memory prophecyClaim = ProphecyClaim(
+                _claimType,
+                _cosmosSender,
+                _ethereumReceiver,
+                msg.sender,
+                tokenAddress,
+                symbol,
+                _amount,
+                Status.Pending
+            );
 
-        // Increment count and add the new ProphecyClaim to the mapping
-        prophecyClaimCount = prophecyClaimCount.add(1);
-        prophecyClaims[prophecyClaimCount] = prophecyClaim;
-
-        emit LogNewProphecyClaim(
-            prophecyClaimCount,
-            _claimType,
-            _cosmosSender,
-            _ethereumReceiver,
-            msg.sender,
-            tokenAddress,
-            symbol,
-            _amount
-        );
+            // Increment count and add the new ProphecyClaim to the mapping
+            prophecyClaimCount = prophecyClaimCount.add(1);
+            prophecyClaims[_prophecyID] = prophecyClaim;
```

Comment:
howmuch extra gas cost does adding this and the keccak256 hashing add to the transaction? and howmuch in total assuming 5 validators does it add to the full relay?

Reply:
I tested the tx with and without keccak256. The gas with keccak256 is 1455766, including the create ERC 20 contract and mint token to recipient. gas without keccak256 is 1455092. the cost of keccak256 is about 674.
For one validator, the gas depends on several factors:
1. if it is a first lock for a new cosmos token, then validator need pay the gas for create contract
2. if it is first tx for the prophecy, the validator need pay gas for create the prophecy
3. if it is the one to get threshold, the validator need pay gas for mint token to recipient.

the total cost for 5 validators may vary, I test it several times. 
the average for one validator is about 350,000 with keccak256.
So total is around 1,750,000 for five validators.

Reply:
By cutting out the separate transaction to create an oracle claim after the prophecy has been created, we are saving gas there as well.